[pytorch] split out trace type generator and migrate to new codegen model

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47438 [pytorch] split out trace type generator and migrate to new codegen model**
* #47437 [pytorch] factor out return_names api
* #47436 [pytorch] TraceType codegen tweak - newline before redispatch call

Differential Revision: [D24808211](https://our.internmc.facebook.com/intern/diff/D24808211)

ljk53 committed:
[pytorch] split out trace type generator and migrate to new codegen model

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] split out trace type generator and migrate to new codegen model"




[ghstack-poisoned]

bhosmer reviewed (approved):
This is a nice chunk of work - will be much easier to rationalize the codegen itself now that it's based on the new data model and moved to its own module. The inline comments are all minor - given byte for byte compatibility, I don't think there's much reason to do heavy rationalization of the codegen logic itself yet (beyond the direct benefits of the data model, that is), since most nontrivial moves would definitely touch the output.

ljk53 committed:
Update on "[pytorch] split out trace type generator and migrate to new codegen model"




[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] split out trace type generator and migrate to new codegen model"


Differential Revision: [D24808211](https://our.internmc.facebook.com/intern/diff/D24808211)

[ghstack-poisoned]

ljk53 commented on tools/autograd/gen_trace_type.py:
```diff
@@ -0,0 +1,412 @@
+        return False
+    name = cpp.name(f.func)
+    base_name = f.func.name.name.base
+    if base_name in DONT_RECORD_TRACE or name in DONT_RECORD_TRACE:
```
changed to check `base_name` (also changed `_coalesced_` to `_coalesced`)

ljk53 commented on tools/autograd/gen_trace_type.py:
```diff
@@ -0,0 +1,412 @@
+    outplace_trace_name = f.func.name.name.base
+    inplace_trace_name = cpp.name(f.func)
+    outplace_trace_name = RENAME_TRACE.get(outplace_trace_name, outplace_trace_name)
+    inplace_trace_name = RENAME_TRACE.get(inplace_trace_name, inplace_trace_name)
```
You mean both `outplace_trace_name` and `inplace_trace_name` or just `inplace_trace_name`? Only `inplace_trace_name` never matches `RENAME_TRACE` - `outplace_trace_name` / `trace_name` above still can.

TBH I'm not 100% sure why the old code names it `outplace_trace_name` - it's used on the `if (force_outplace)` branch below but it's clearly derived from the base name (the old code called utils.uninplace_api_name which removes `_out` suffix manually).

ljk53 commented on tools/codegen/gen.py:
```diff
@@ -821,9 +821,10 @@ def write_with_template(self, filename: str, template_fn: str,
-                comment = "@" + "generated by aten/src/ATen/gen.py"
-                comment += " from {}".format(os.path.basename(template_fn))
-                env['generated_comment'] = comment
+                if 'generated_comment' not in env:
```
donno - @ezyang what does 'correct location' mean? full path?

ljk53 commented on tools/autograd/gen_autograd.py:
```diff
@@ -243,6 +243,10 @@ def is_operator_selected_for_training(decl):
         from .gen_variable_type import gen_variable_type
         gen_variable_type(out, aten_decls, template_path)
 
+        from . import gen_trace_type
```
Where can I learn more about the `==` perf regression? Does the lazy import only regress the first run (i.e. benchmark measure problem) or does it actually regress every run (poor implementation of lazy import of python)?
Lazy import might have the benefit of not paying all the cost at cold start time - I don't know if it's desired in some cases for the python frontend code. (hmm, I agree even if so people probably didn't do so deliberately to hide code start latency - they might simply want to localize the imports :-) )

Any way, the codegen script only runs at build time and this is not on hot path so I don't think perf matters here. Unless we want to enforce the top-level import throughout the python codebase for the eager mode frontend code where perf matters.

ljk53 committed:
Update on "[pytorch] split out trace type generator and migrate to new codegen model"


Differential Revision: [D24808211](https://our.internmc.facebook.com/intern/diff/D24808211)

[ghstack-poisoned]

ezyang commented:
just added @bdhirsh for visibility, since Brian's been looking at this code recently
